File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/go-counting/go-counting.js
const OWNER = {
  NONE: 'NONE',
  BLACK: 'BLACK',
  WHITE: 'WHITE'
};

export class GoCounting {
  constructor(board) {
    this.board = board.map(row => row.split(''));
    this.rows = this.board.length;
    this.cols = this.rows > 0 ? this.board[0].length : 0;
  }

  getTerritory(x, y) {
    if (x < 0 || x >= this.rows || y < 0 || y >= this.cols) {
      return { error: 'Invalid coordinate' };
    }

    if (this.board[x][y] !== ' ') {
      return { owner: OWNER.NONE, territory: [] };
    }

    const visited = new Set();
    const territory = [];
    let borders = new Set();
    
    const stack = [[x, y]];
    
    while (stack.length > 0) {
      const [cx, cy] = stack.pop();
      const key = `${cx},${cy}`;
      
      if (visited.has(key)) continue;
      visited.add(key);
      
      if (this.board[cx][cy] !== ' ') continue;
      
      territory.push([cx, cy]);
      
      // Check all 4 neighbors
      for (const [nx, ny] of this.getNeighbors(cx, cy)) {
        const cell = this.board[nx][ny];
        if (cell === 'B' || cell === 'W') {
          borders.add(cell);
        } else if (cell === ' ') {
          stack.push([nx, ny]);
        }
      }
    }
    
    // Sort territory coordinates
    territory.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    
    // Determine owner based on borders
    let owner = OWNER.NONE;
    if (borders.size === 1) {
      owner = borders.has('B') ? OWNER.BLACK : OWNER.WHITE;
    }
    
    return { owner, territory };
  }

  getTerritories() {
    const territories = {
      territoryBlack: [],
      territoryWhite: [],
      territoryNone: []
    };
    
    const visited = new Set();
    
    for (let x = 0; x < this.rows; x++) {
      for (let y = 0; y < this.cols; y++) {
        if (this.board[x][y] === ' ' && !visited.has(`${x},${y}`)) {
          const territory = this.getTerritory(x, y);
          if (territory.owner === OWNER.BLACK) {
            territories.territoryBlack.push(...territory.territory);
          } else if (territory.owner === OWNER.WHITE) {
            territories.territoryWhite.push(...territory.territory);
          } else {
            territories.territoryNone.push(...territory.territory);
          }
          territory.territory.forEach(([tx, ty]) => visited.add(`${tx},${ty}`));
        }
      }
    }
    
    // Sort all territories
    territories.territoryBlack.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    territories.territoryWhite.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    territories.territoryNone.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    
    return territories;
  }

  getNeighbors(x, y) {
    const neighbors = [];
    if (x > 0) neighbors.push([x - 1, y]);
    if (x < this.rows - 1) neighbors.push([x + 1, y]);
    if (y > 0) neighbors.push([x, y - 1]);
    if (y < this.cols - 1) neighbors.push([x, y + 1]);
    return neighbors;
  }
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/go-counting/src/main/java/GoCounting.java
import java.awt.Point;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class GoCounting {
    private enum Cell { BLACK, WHITE, EMPTY }
    private Cell[][] board;
    private int width;
    private int height;

    GoCounting(String board) {
        String[] lines = board.split("\n");
        this.height = lines.length;
        this.width = lines[0].length();
        this.board = new Cell[height][width];
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                char c = lines[y].charAt(x);
                switch (c) {
                    case 'B': this.board[y][x] = Cell.BLACK; break;
                    case 'W': this.board[y][x] = Cell.WHITE; break;
                    default: this.board[y][x] = Cell.EMPTY; break;
                }
            }
        }
    }

    Player getTerritoryOwner(int x, int y) {
        if (!isValid(x, y)) {
            throw new IllegalArgumentException("Invalid coordinate");
        }
        if (board[y][x] != Cell.EMPTY) {
            return Player.NONE;
        }
        
        Set<Point> territory = new HashSet<>();
        Set<Player> borders = new HashSet<>();
        floodFill(x, y, territory, borders);
        
        if (borders.size() == 1) {
            return borders.iterator().next();
        }
        return Player.NONE;
    }

    Set<Point> getTerritory(int x, int y) {
        if (!isValid(x, y)) {
            throw new IllegalArgumentException("Invalid coordinate");
        }
        if (board[y][x] != Cell.EMPTY) {
            return Set.of();
        }
        
        Set<Point> territory = new HashSet<>();
        Set<Player> borders = new HashSet<>();
        floodFill(x, y, territory, borders);
        return territory;
    }

    Map<Player, Set<Point>> getTerritories() {
        Map<Player, Set<Point>> territories = new HashMap<>();
        territories.put(Player.BLACK, new HashSet<>());
        territories.put(Player.WHITE, new HashSet<>());
        territories.put(Player.NONE, new HashSet<>());
        
        boolean[][] visited = new boolean[height][width];
        
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                if (board[y][x] == Cell.EMPTY && !visited[y][x]) {
                    Set<Point> territory = new HashSet<>();
                    Set<Player> borders = new HashSet<>();
                    floodFill(x, y, territory, borders);
                    
                    Player owner = borders.size() == 1 ? borders.iterator().next() : Player.NONE;
                    territories.get(owner).addAll(territory);
                    
                    for (Point p : territory) {
                        visited[p.y][p.x] = true;
                    }
                }
            }
        }
        return territories;
    }

    private void floodFill(int x, int y, Set<Point> territory, Set<Player> borders) {
        // Use a stack-based approach instead of recursion
        int[] stackX = new int[width * height];
        int[] stackY = new int[width * height];
        int stackPtr = 0;
        
        stackX[stackPtr] = x;
        stackY[stackPtr] = y;
        stackPtr++;
        
        while (stackPtr > 0) {
            stackPtr--;
            int cx = stackX[stackPtr];
            int cy = stackY[stackPtr];
            
            if (!isValid(cx, cy) || board[cy][cx] != Cell.EMPTY || territory.contains(new Point(cx, cy))) {
                continue;
            }
            
            territory.add(new Point(cx, cy));
            
            // Check neighbors
            if (cx > 0) {
                if (board[cy][cx-1] == Cell.BLACK) borders.add(Player.BLACK);
                else if (board[cy][cx-1] == Cell.WHITE) borders.add(Player.WHITE);
                else {
                    stackX[stackPtr] = cx-1;
                    stackY[stackPtr] = cy;
                    stackPtr++;
                }
            }
            if (cx < width-1) {
                if (board[cy][cx+1] == Cell.BLACK) borders.add(Player.BLACK);
                else if (board[cy][cx+1] == Cell.WHITE) borders.add(Player.WHITE);
                else {
                    stackX[stackPtr] = cx+1;
                    stackY[stackPtr] = cy;
                    stackPtr++;
                }
            }
            if (cy > 0) {
                if (board[cy-1][cx] == Cell.BLACK) borders.add(Player.BLACK);
                else if (board[cy-1][cx] == Cell.WHITE) borders.add(Player.WHITE);
                else {
                    stackX[stackPtr] = cx;
                    stackY[stackPtr] = cy-1;
                    stackPtr++;
                }
            }
            if (cy < height-1) {
                if (board[cy+1][cx] == Cell.BLACK) borders.add(Player.BLACK);
                else if (board[cy+1][cx] == Cell.WHITE) borders.add(Player.WHITE);
                else {
                    stackX[stackPtr] = cx;
                    stackY[stackPtr] = cy+1;
                    stackPtr++;
                }
            }
        }
    }

    private boolean isValid(int x, int y) {
        return x >= 0 && x < width && y >= 0 && y < height;
    }
}
